# Importing Libraries

In [2]:
# import libraries
import pandas as pd
import numpy as np
import os

# Importing Data

## Defining Path Variable

In [3]:
# defining path
path = r'C:\Users\khand\Desktop\Careerfoundry\Data Immersion\Achievement - 4\04-2022 Instacart Basket Analysis'

## Importing "orders.csv"

In [4]:
# importing "orders.csv"
df_ords = pd.read_csv(os.path.join(path, '2. Data', 'Original Data', 'orders.csv'), index_col = False)

## Importing "products.csv"

In [5]:
# importing "products.csv"
df_prods = pd.read_csv(os.path.join(path, '2. Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
# Top 5 rows of df_ords
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [7]:
# Deleting the unwanted columns from dataframe
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [8]:
# frequency distribution of the 'days_since_prior_order' column
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

## Overwriting the dataframe permanently

In [9]:
# Permanently deleting the unwanted column from the dataset
df_ords = df_ords.drop(columns = ['eval_set'])

## Renaming the clumn 'order_dow' to 'order_day_of_week'

In [10]:
# renaming a column
df_ords.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [11]:
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## Changing data type of 'order_id' column to 'str'

In [12]:
# Changing datatype of a column
df_ords['order_id'] = df_ords['order_id'].astype('str')

### Checking data type of the 'orders_id' column

In [13]:
# Checking data type
df_ords['order_id'].dtype

dtype('O')

# Importing 'departments.csv'

In [14]:
# Importing the "departments.csv" dataframe
df_dep = pd.read_csv(os.path.join(path, '2. Data', 'Original Data', 'departments.csv'))

In [15]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

## Creating a data dictionary

### Transposing the 'df_dep' data frame

In [16]:
# Transposing the dataframe
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## Permanently overwriting the 'df_dep' dataframe in long format

In [17]:
# Permanentry transposing the dataframe
df_dep = df_dep.T

In [18]:
df_dep.head()

0
department_id  department
1                  frozen
2                   other
3                  bakery
4                 produce

## Adding index to dataframe 'df_data'

In [19]:
Adding index to the dataframe
df_dep.reset_index()

SyntaxError: invalid syntax (Temp/ipykernel_20004/3153433789.py, line 1)

### Assigning the first row as value of variable "new_header" 

In [20]:
# Assigning the values in the first row to the "new_header" variable
new_header = df_dep.iloc[0]

In [21]:
new_header

0    department
Name: department_id, dtype: object

### Turning the row 1 and beyond in to new dataframe

In [22]:
# turning the records in row 1 and beyond into a new dataframe
df_dep_t_new = df_dep[1:]

In [23]:
df_dep_t_new.head()

0
1   frozen
2    other
3   bakery
4  produce
5  alcohol

### Assigning the values of "new_header" variable as the column names of the new database

In [24]:
# Assigning the values of the "new_header" variable as column headers
df_dep_t_new.columns = new_header

In [25]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

### Turning the new dataframe into a data dictionary

In [26]:
# Turing the dataframe into a data dictionary
data_dict = df_dep_t_new.to_dict('index')

In [27]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

## Looking for records with "department_id" of 19

In [28]:
# Searching for records with department id: 19
df_prods['department_id'] == 19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

## Isolating only the records with "department_id" of 19

In [29]:
# filtering records with department id: 19
df_prods[df_prods['department_id'] == 19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

## Turning the filtered portion of the dataframe into a new dataframe

In [30]:
# turning the filtered records into a new dataframe
df_snacks = df_prods[df_prods['department_id'] == 19]

In [31]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

## Using the loc function to filter

In [32]:
# Using the loc function to filter
df_prods.loc[df_prods['department_id'] == 19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

## Using isin([]) to filter with multiple values/list

In [33]:
# Using isin([]) to filter
df_prods.loc[df_prods['department_id'].isin([19])]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

# Exporting dataframes

In [34]:
df_ords.to_csv(os.path.join(path, '2. Data', 'Prepared Data', 'Orders_prepared.csv'), index = None)

In [35]:
df_snacks.to_csv(os.path.join(path, '2. Data', 'Prepared Data', 'Snacks.csv'), index = None)

In [36]:
df_dep_t_new.to_csv(os.path.join(path, '2. Data', 'Prepared Data', 'Data_dictionary.csv'), index = None)

# Task 4.4

In [37]:
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329        1             1                  2                  8   
1  2398795        1             2                  3                  7   
2   473747        1             3                  3                 12   
3  2254736        1             4                  4                  7   
4   431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## Changing data types of Identifier variables

In [38]:
# Changing data type of "order_id" variable to "string"
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [39]:
# Changing data type of "user_id" variable to "sting"
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [40]:
# Checking data type of "order_id" variable
df_ords['order_id'].dtype

dtype('O')

In [41]:
# Checking data type of "user_id" variable
df_ords['user_id'].dtype

dtype('O')

## Changing column name

In [42]:
# Changing the name of the "order_hour_of_day" column
df_ords.rename(columns = {'order_hour_of_day' : 'order_time_24h_format'}, inplace = True)

In [43]:
df_ords.head()

order_id user_id  order_number  order_day_of_week  order_time_24h_format  \
0  2539329       1             1                  2                      8   
1  2398795       1             2                  3                      7   
2   473747       1             3                  3                     12   
3  2254736       1             4                  4                      7   
4   431534       1             5                  4                     15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## Finding the busiest hour of the day

In [44]:
# Finding the frequency of the values in the "order_time_hour_of_day" 
df_ords['order_time_24h_format'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_time_24h_format, dtype: int64

#### The value 10 has the highest frequency, meaning 10 am is the busiest hour of the day in terms of order numbers 

## Determining the meaning of the value '4' in the "department_id" column

In [45]:
# Printing the value corresponding to the department id: "4"
print(data_dict.get('4'))

{'department': 'produce'}


## Subsetting order data from only breakfast department

In [46]:
#Printing the whole data dictionary
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

#### "breakfast" items are represented by the value "14" in the "department_id" column

### Subsetting the "breakfast" data

In [47]:
# Filtering only the records with "department_id" of "14" from the "df_prods" dataframe
df_prods.loc[df_prods['department_id'].isin([14])]

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [48]:
# Turning the filtered "breakfast" department data in to new dataframe
df_breakfast = df_prods.loc[df_prods['department_id'].isin([14])]

In [49]:
# Finding the number of rows and columns in the dataframe
df_breakfast.shape

(1116, 5)

#### The "df_breakfast" dataframe has 1116 records.

### Subsetting the "dinner party" data

In [50]:
# Filtering the records with department_id 5, 7, 12, 20 which refer to "alchohol", "beverage", "meat/seafood" and "deli"
df_prods.loc[df_prods['department_id'].isin([5, 7, 12, 20])]

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

In [51]:
# Turining the filtered data in to a new dataframe
df_dinner_party = df_prods.loc[df_prods['department_id'].isin([5, 7, 12, 20])]

In [52]:
# finding the numner of rows and columns in the "df_dinner_party" dataset
df_dinner_party.shape

(7650, 5)

#### The "df_dinner_party" dataframe has 7650 records

### Subsetting records that have "1" as value in the "user_id" column.

In [53]:
# Filtering the records with user id: 1
df_ords.loc[df_ords['user_id'].isin(['1'])]

order_id user_id  order_number  order_day_of_week  order_time_24h_format  \
0   2539329       1             1                  2                      8   
1   2398795       1             2                  3                      7   
2    473747       1             3                  3                     12   
3   2254736       1             4                  4                      7   
4    431534       1             5                  4                     15   
5   3367565       1             6                  2                      7   
6    550135       1             7                  1                      9   
7   3108588       1             8                  1                     14   
8   2295261       1             9                  1                     16   
9   2550362       1            10                  4                      8   
10  1187899       1            11                  4                      8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [54]:
# Turning the filtered records in to a separate dataframe
df_user_id_1 = df_ords.loc[df_ords['user_id'].isin(['1'])]

### Basic statistics of the records under user id: 1

In [55]:
# Number of rows and columns in the "df_user_id_1" dataframe
df_user_id_1.shape

(11, 6)

In [56]:
# Descriptive analysis of the "df_user_id_1" dataframe
df_user_id_1.describe()

order_number  order_day_of_week  order_time_24h_format  \
count     11.000000          11.000000              11.000000   
mean       6.000000           2.636364              10.090909   
std        3.316625           1.286291               3.477198   
min        1.000000           1.000000               7.000000   
25%        3.500000           1.500000               7.500000   
50%        6.000000           3.000000               8.000000   
75%        8.500000           4.000000              13.000000   
max       11.000000           4.000000              16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

## Exporting dataframes

In [57]:
# Exporting dataframe "df_ords"
df_ords.to_csv(os.path.join(path, '2. Data', 'Prepared Data', 'orders_wrangled.csv'), index = None)


In [58]:
# Exporting dataframe "df_dep_t_new"
df_dep_t_new.to_csv(os.path.join(path, '2. Data', 'Prepared Data', 'departments_wrangled.csv'), index = None)